File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int>s;
        unordered_map<int, int>mp;
        //dry run the following code
        for(int n:nums2){
            while(!s.empty() and s.top() < n){
                mp[s.top()] = n;
                s.pop();
            }
            s.push(n);
        }
        vector<int>ans;
        for(auto n : nums1){
            ans.push_back(mp.count(n) ? mp[n] : -1);
            
        }
        return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        #making a stack and a map
        s = []
        mp = {}
        #for checking the nums2
        for n in nums2:
            while s and s[-1] < n: #while s is not empty and the top of s < n
                mp[s[-1]] = n
                s.pop()
            s.append(n)
        ans = []
        for n in nums1:
            ans.append(mp[n] if n in mp else -1)
        return ans

        